plane/src/typed_unix_socket/server.rs

```diff
@@ -0,0 +1,220 @@
+    for TypedUnixSocketServer<MessageToServer, MessageToClient>
+where
+    MessageToServer: Send + Sync + 'static + Clone + Debug + Serialize + for<'de> Deserialize<'de>,
+    MessageToClient: Send + Sync + 'static + Clone + Debug + Serialize + for<'de> Deserialize<'de>,
+{
+    fn drop(&mut self) {
+        if let Err(e) = fs::remove_file(&self.socket_path) {
+            tracing::warn!("Failed to remove socket file: {}", e);
+        }
+    }
+}
+
+async fn handle_connection<MessageToServer, MessageToClient>(
+    stream: UnixStream,
+    event_tx: broadcast::Sender<MessageToServer>,
+    request_tx: broadcast::Sender<WrappedMessage<MessageToServer>>,
+    mut response_rx: broadcast::Receiver<WrappedMessage<MessageToClient>>,
+) -> Result<(), Box<dyn std::error::Error + Send + Sync>>
+where
+    MessageToServer: Send + Sync + 'static + Clone + Debug + Serialize + for<'de> Deserialize<'de>,
+    MessageToClient: Send + Sync + 'static + Clone + Debug + Serialize + for<'de> Deserialize<'de>,
+{
+    let (read_half, write_half) = tokio::io::split(stream);
+    let reader = BufReader::new(read_half);
+    let writer = BufWriter::new(write_half);
+
+    let mut lines = reader.lines();
+    let mut writer = writer;
+
+    // Task to handle receiving messages
+    let recv_task = tokio::spawn({
+        let event_tx = event_tx.clone();
+        async move {
+            loop {
+                match lines.next_line().await {
+                    Ok(Some(line)) => {
+                        let msg: WrappedMessage<MessageToServer> = match serde_json::from_str(&line)
+                        {
+                            Ok(msg) => msg,
+                            Err(e) => {
+                                tracing::error!("Error deserializing message: {}", e);
+                                continue;
+                            }
+                        };
+                        match msg {
+                            WrappedMessage { id: Some(_), .. } => {
+                                if let Err(e) = request_tx.send(msg) {
+                                    tracing::error!("Error sending request: {}", e);
+                                }
+                            }
+                            WrappedMessage { id: None, message } => {
+                                if let Err(e) = event_tx.send(message) {
+                                    tracing::error!("Error sending event: {}", e);
+                                }
+                            }
+                        }
+                    }
+                    Ok(None) => {
+                        tracing::error!("Connection closed by client");
+                    }
+                    Err(e) => {
+                        tracing::error!("Error reading line: {}", e);
+                    }
+                }
+            }
+        }
+    });
+
+    // Task to handle sending responses
+    let send_task = tokio::spawn(async move {
+        loop {
```

Comment:
I think this loop never stops, even if the connection has reconnected?

I am tempted to just break from the loop if we encounter an error; reconnecting in that case doesn't sound like a bad idea even if it may have been recoverable.